File: src/main/java/minic/ui/MiniCLexerOverlayModelFactory.java
package minic.ui;

import minic.uiapi.UiLexerTokenVisualDto;
import minic.uiapi.UiStageVisualDto;

import java.util.ArrayList;
import java.util.List;

/**
 * 根据 Lexer visual DTO 生成等宽源码遮罩模型。
 */
public final class MiniCLexerOverlayModelFactory {
    /**
     * 创建 lexer 遮罩行模型。
     *
     * @param source 源码文本
     * @param visual 当前阶段 visual DTO
     * @return 行模型
     */
    public List<MiniCLexerOverlayLine> create(String source, UiStageVisualDto visual) {
        UiLexerTokenVisualDto activeToken = visual.lexerTokens().stream()
                .filter(UiLexerTokenVisualDto::active)
                .findFirst()
                .orElse(null);
        ArrayList<MiniCLexerOverlayLine> rows = new ArrayList<>();
        int lineStart = 0;
        int lineNumber = 1;
        for (int offset = 0; offset <= source.length(); offset++) {
            if (offset == source.length() || source.charAt(offset) == '\n') {
                rows.add(new MiniCLexerOverlayLine(lineNumber, segments(source, lineStart, offset, activeToken)));
                lineNumber++;
                lineStart = offset + 1;
            }
        }
        if (rows.isEmpty()) {
            rows.add(new MiniCLexerOverlayLine(1, List.of(new MiniCLexerOverlaySegment("", false))));
        }
        return List.copyOf(rows);
    }

    private List<MiniCLexerOverlaySegment> segments(String source, int lineStart, int lineEnd, UiLexerTokenVisualDto activeToken) {
        int activeStart = activeToken == null ? -1 : activeToken.startOffset();
        int activeEnd = activeToken == null ? -1 : activeToken.endOffset();
        if (activeToken == null || activeEnd <= lineStart || activeStart >= lineEnd) {
            return List.of(new MiniCLexerOverlaySegment(source.substring(lineStart, lineEnd), false));
        }

        ArrayList<MiniCLexerOverlaySegment> segments = new ArrayList<>();
        int beforeEnd = Math.max(lineStart, Math.min(activeStart, lineEnd));
        int hotStart = Math.max(lineStart, activeStart);
        int hotEnd = Math.max(hotStart, Math.min(activeEnd, lineEnd));
        if (beforeEnd > lineStart) {
            segments.add(new MiniCLexerOverlaySegment(source.substring(lineStart, beforeEnd), false));
        }
        segments.add(new MiniCLexerOverlaySegment(source.substring(hotStart, hotEnd), true));
        if (lineEnd > hotEnd) {
            segments.add(new MiniCLexerOverlaySegment(source.substring(hotEnd, lineEnd), false));
        }
        return List.copyOf(segments);
    }
}

File: uiweb/src/visual/MiniCLexerOverlayModelFactory.ts
import type { JavaMirrorFile, UiLexerTokenVisualDto, UiStageVisualDto } from "../translation/javaMirror";
import { MiniCLexerOverlayLine } from "./MiniCLexerOverlayLine";
import { MiniCLexerOverlaySegment } from "./MiniCLexerOverlaySegment";

export const miniCLexerOverlayModelFactoryMirror = {
  javaPath: "src/main/java/minic/uilocal/visual/MiniCLexerOverlayModelFactory.java",
  webPath: "uiweb/src/visual/MiniCLexerOverlayModelFactory.ts",
  packageName: "minic.uilocal.visual",
  exportName: "MiniCLexerOverlayModelFactory",
  kind: "class",
  imports: [
    "java.util.ArrayList",
    "java.util.List",
    "minic.uiapi.UiLexerTokenVisualDto",
    "minic.uiapi.UiStageVisualDto"
  ],
  fields: [],
  methods: [
    {
      "name": "create",
      "signature": "create(String source,UiStageVisualDto visual)"
    },
    {
      "name": "segments",
      "signature": "segments(String source,int lineStart,int lineEnd,UiLexerTokenVisualDto activeToken)"
    }
  ],
} as const satisfies JavaMirrorFile;

export class MiniCLexerOverlayModelFactory {
  static readonly mirror = miniCLexerOverlayModelFactoryMirror;

  create(source: string, visual: UiStageVisualDto | null | undefined): readonly MiniCLexerOverlayLine[] {
    const activeToken = visual?.lexerTokens.find((token) => token.active) ?? null;
    const rows: MiniCLexerOverlayLine[] = [];
    let lineStart = 0;
    let lineNumber = 1;

    for (let offset = 0; offset <= source.length; offset++) {
      if (offset === source.length || source.charAt(offset) === "\n") {
        rows.push(new MiniCLexerOverlayLine(lineNumber, this.segments(source, lineStart, offset, activeToken)));
        lineNumber += 1;
        lineStart = offset + 1;
      }
    }

    if (rows.length === 0) {
      rows.push(new MiniCLexerOverlayLine(1, [new MiniCLexerOverlaySegment("", false)]));
    }
    return rows;
  }

  private segments(
    source: string,
    lineStart: number,
    lineEnd: number,
    activeToken: UiLexerTokenVisualDto | null,
  ): readonly MiniCLexerOverlaySegment[] {
    const activeStart = activeToken?.startOffset ?? -1;
    const activeEnd = activeToken?.endOffset ?? -1;
    if (!activeToken || activeEnd <= lineStart || activeStart >= lineEnd) {
      return [new MiniCLexerOverlaySegment(source.slice(lineStart, lineEnd), false)];
    }

    const segments: MiniCLexerOverlaySegment[] = [];
    const beforeEnd = Math.max(lineStart, Math.min(activeStart, lineEnd));
    const hotStart = Math.max(lineStart, activeStart);
    const hotEnd = Math.max(hotStart, Math.min(activeEnd, lineEnd));
    if (beforeEnd > lineStart) {
      segments.push(new MiniCLexerOverlaySegment(source.slice(lineStart, beforeEnd), false));
    }
    segments.push(new MiniCLexerOverlaySegment(source.slice(hotStart, hotEnd), true));
    if (lineEnd > hotEnd) {
      segments.push(new MiniCLexerOverlaySegment(source.slice(hotEnd, lineEnd), false));
    }
    return segments;
  }
}

export default MiniCLexerOverlayModelFactory;
